WCSimfitQunHitPreprocess.c

```diff
@@ -0,0 +1,368 @@
+        fitQunDir[1] = fq1rdir[0][2][2];   //g
+        fitQunDir[2] = fq1rdir[0][2][1];   //g
+
+        //extrapolate the fitQun vertex to obtain the entrance and exit position (fitQunEntrance[3] and fitQunExit[3])
+        if (extrapolation(fitQunPos, fitQunDir, fitQunEntrance, fitQunExit)) {
+            ignored_track++;
+            continue;
+        }   //g for fitQunEntrance[3] and fitQunExit[3]
+
+        //selection on the main track by the fitQun entrance and exit position
+        if (selection(fitQunEntrance, fitQunExit)) {
+            ignored_track++;
+            continue;
+        }
+        fitQunVerTime = fq1rt0[0][2];   //g
+
+        wt->GetEntry(i);
+        WCSimRootTrigger* wcsimrootevent = wcsimrootsuperevent->GetTrigger(0);
+        //search for main track
+        int ntrack = wcsimrootevent->GetNtrack_slots();
+        for (int itrack=0; itrack<ntrack; itrack++) {
+            wcsimroottrack = dynamic_cast<WCSimRootTrack*> (wcsimrootevent->GetTracks()->At(itrack));
+            if (wcsimroottrack->GetId()==1) {break;}
+        }
+        int nCrossing = wcsimroottrack->GetBoundaryPoints().size();
+        if (nCrossing!=4) {
+            ignored_track++;
+            continue;
+        } //ignoring tracks that did not enter or terminated in the tank
+
+        //calculate WCSim entrance and exit points
+        WCSimEntrance[0] = -wcsimroottrack->GetBoundaryPoints().at(1).at(0)/10;    //g
+        WCSimEntrance[1] = wcsimroottrack->GetBoundaryPoints().at(1).at(2)/10;    //g
+        WCSimEntrance[2] = wcsimroottrack->GetBoundaryPoints().at(1).at(1)/10;    //g
+        WCSimExit[0] = -wcsimroottrack->GetBoundaryPoints().at(2).at(0)/10;    //g
+        WCSimExit[1] = wcsimroottrack->GetBoundaryPoints().at(2).at(2)/10;    //g
+        WCSimExit[2] = wcsimroottrack->GetBoundaryPoints().at(2).at(1)/10;    //g
+
+        //calculate the WCSim travel direction
+        WCSimDir[0] = WCSimExit[0] - WCSimEntrance[0];
+        WCSimDir[1] = WCSimExit[1] - WCSimEntrance[1];
+        WCSimDir[2] = WCSimExit[2] - WCSimEntrance[2];
+        double WCSimDist = sqrt(WCSimDir[0]*WCSimDir[0]+WCSimDir[1]*WCSimDir[1]+WCSimDir[2]*WCSimDir[2]);
+        WCSimDir[0] = WCSimDir[0]/WCSimDist;    //g
+        WCSimDir[1] = WCSimDir[1]/WCSimDist;    //g
+        WCSimDir[2] = WCSimDir[2]/WCSimDist;    //g
+        WCSimStartPos[0] = -wcsimroottrack->GetStart(0);    //g
+        WCSimStartPos[1] = wcsimroottrack->GetStart(2);    //g
+        WCSimStartPos[2] = wcsimroottrack->GetStart(1);    //g
+        THNum = wcsimrootevent->GetNcherenkovhittimes();
+
+        TClonesArray* WRCDHArray = wcsimrootevent->GetCherenkovDigiHits();    //Array of WCSimRootCherenkovDigiHit pointer
+        DHNum = WRCDHArray->GetEntriesFast();
+        for (int j=0; j<DHNum; j++) {
+            //ignore extremely long digit time (>200 ns)
+            if (((WCSimRootCherenkovDigiHit*) WRCDHArray->At(j))->GetT()>200) {
+                WRCDHArray->RemoveAt(j);
```

Comment:
since you don't need this array to do selection for true hits below, just fill the `DHTree` here,no need to remove those entry.